Regression Fix + Refine Status Updates + Resolve Edge Cases + Code Cleanup

sakshibobade21 opened the pull request:
<!-- Thank you for submitting a PR to Zowe! To help us understand, test, and give feedback on your code, please fill in the details below. -->

## Proposed changes
<!-- Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue. -->

This PR addresses Issue: 
This PR resolves a regression issue where the status for the VSAM stage was being updated without the stage being completed. It also addresses edge cases where the status was not updated as expected when stages were skipped or completed. The implementation now ensures that the status of stages and substages is properly stored and initialized accordingly. Additionally, this PR includes some code cleanup



This PR depends upon the following PRs:

## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Change in a documentation
- [ ] Refactor the code 
- [ ] Chore, repository cleanup, updates the dependencies.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## PR Checklist
Please delete options that are not relevant.
- [ ] If the changes in this PR are meant for the next release / mainline, this PR targets a "staging" branch.
- [ ] My code follows the style guidelines of this project (see: [Contributing guideline](https://github.com/zowe/zen/blob/v2.x/staging/CONTRIBUTING.md))
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] New and existing unit tests pass locally with my changes
- [ ] video or image is included if visual changes are made
<!-- - [ ] Relevant update to CHANGELOG.md -->
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works, or describe a test method below

## Testing
<!-- Describe how this code should be tested. I've you've added an automated or unit test, then describe how to run it. Otherwise, describe how you have tested it and how others should test it. -->

## Further comments
<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, or if there are follow-up tasks and TODOs etc... -->

sakshibobade21 committed:
Updating the stepper

sakshibobade21 committed:
Passthrough

sakshibobade21 committed:
Merge remote-tracking branch 'origin' into regression/vsam-status

timgerstel committed:
Merge branch 'v2.x/staging' into regression/vsam-status

timgerstel commented on src/renderer/components/common/Stepper.tsx:
```diff
@@ -176,7 +176,7 @@ export default function HorizontalLinearStepper({stages, initialization}:{stages
   const getStepIcon = (error: any, stageId: number, isSubStep?: boolean, subStepId?: number) => {
     
-    if (!error || (isSubStep && getProgress(stages[stageId].subStages[subStepId].statusKey)) || (!isSubStep && getProgress(stages[stageId].statusKey))) {
+    if ((isSubStep && getProgress(stages[stageId].subStages[subStepId].statusKey)) || (!isSubStep && getProgress(stages[stageId].statusKey))) {
```
Wait why dont we want to check for error

sakshibobade21 replied:
We no longer need the "error" prop. We show the warning icon depending on the skip status of the stage.

timgerstel replied:
Should probably remove the arg then if its not needed

sakshibobade21 replied:
Yes it is removed.

sakshibobade21 committed:
regression fix

sakshibobade21 committed:
Merge branch 'regression/vsam-status' of https://github.com/zowe/zen into regression/vsam-status

sakshibobade21 committed:
Updating the stepper icon logic

sakshibobade21 committed:
Removing the space

sakshibobade21 committed:
Updating the interface name

sakshibobade21 committed:
indendations

sakshibobade21 committed:
Updating the stepper

sakshibobade21 committed:
Updating the dataset instllation

sakshibobade21 committed:
Removing the eventlisterner

sakshibobade21 committed:
Handling edge cases for instllation

sakshibobade21 committed:
code cleanup

sakshibobade21 committed:
Updating the substep skip array

sakshibobade21 committed:
Removing error prop

timgerstel commented on src/renderer/components/stages/progress/StageProgressStatus.ts:
```diff
@@ -274,19 +291,58 @@ export const mapAndGetSkipStatus = (subStageId: number): boolean => {
 
-export const setSubStageSkipStatus = (key: keyof SkipState, newValue: boolean): void => {
+export const mapAndSetStepSkipStatus = (stageId: number, value: boolean): void => {
+  setStageSkipStatus(stepSkipKeysArray[stageId-1], value);
```
this needs a check to make sure stageId isn't an out of bounds index

sakshibobade21 replied:
Done. Thanks

timgerstel commented on src/renderer/components/stages/progress/StageProgressStatus.ts:
```diff
@@ -274,19 +291,58 @@ export const mapAndGetSkipStatus = (subStageId: number): boolean => {
+    skipStatus.reviewInstallation
+  ]
+
+  return skipStatusArray[subStageId-1];
```
check if array index out of bounds

sakshibobade21 replied:
Done. Thanks

sakshibobade21 committed:
Removing error attr

sakshibobade21 committed:
Adding some coditions

sakshibobade21 committed:
Updating the consitions

sakshibobade21 committed:
Merge remote-tracking branch 'origin' into regression/vsam-status

sakshibobade21 committed:
Bugfix

sakshibobade21 committed:
Updating progress in edge cases

sakshibobade21 committed:
Fixing the status of the networking and the launch config stage

sakshibobade21 committed:
merging conflicts

DivergentEuropeans commented on src/renderer/components/stages/Security.tsx:
```diff
@@ -187,8 +186,6 @@ const Security = () => {
       clearInterval(timer);
       securityProceedActions(false);
-      stages[STAGE_ID].subStages[SUB_STAGE_ID].isSkipped = true;
-      stages[STAGE_ID].isSkipped = true;
```
Can you explain these changes a bit more? So why are we removing some of these skipped statuses when things fail? If they are green, and are re-run and they fail, status should go from green to skipped no? (well ideally "error" status but it's nbd)

sakshibobade21 replied:
If things fail, we call updateProgress(false). Update progress internally calls setStageSkipStatus(true) which sets the status for the stages.

sakshibobade21 replied:
This changes were made to cleanup the code and remove redundant lines of code

DivergentEuropeans commented on src/renderer/components/stages/Networking.tsx:
```diff
@@ -569,8 +569,6 @@ const Networking = () => {
     dispatch(setNextStepEnabled(getProgress('networkingStatus')));
     dispatch(setInitializationStatus(isInitComplete()));
-    stages[STAGE_ID].subStages[SUB_STAGE_ID].isSkipped = false;
-    stages[STAGE_ID].isSkipped = isInitializationSkipped;
```
My comment applies here too

sakshibobade21 replied:
If things fail, we call updateProgress(false). Update progress internally calls setStageSkipStatus(true) which sets the status for the stages.

This changes were made to cleanup the code and remove redundant lines of code

DivergentEuropeans replied:
Where do we call it near this code? Not called in this file at least

sakshibobade21 replied:
I removed those lines of code because every time the networking component loads, it incorrectly sets the `skip` status to false. If the networking stage is skipped and the user revisits it, the status should be skipped

sakshibobade21 replied:
Same applies for the launch config stage

DivergentEuropeans commented on src/renderer/components/stages/InitApfAuth.tsx:
```diff
@@ -194,8 +192,6 @@ const InitApfAuth = () => {
         // TODO: Test this
         //alertEmitter.emit('showAlert', err.toString(), 'error');
-        stages[STAGE_ID].subStages[SUB_STAGE_ID].isSkipped = true;
-        stages[STAGE_ID].isSkipped = true;
```
My comment applies here too

DivergentEuropeans commented on src/renderer/components/stages/installation/Installation.tsx:
```diff
@@ -269,21 +270,14 @@ const Installation = () => {
         updateProgress(false);
-        installProceedActions(false);
-        stages[STAGE_ID].subStages[SUB_STAGE_ID].isSkipped = true;
-        stages[STAGE_ID].isSkipped = true;
```
Here too, etc. etc.

DivergentEuropeans replied:
Also did you mean to remove 'installProceedActions' too?

sakshibobade21 replied:
Removed `installProceedActions()` because updateProgress() is calling it internally.

Again this change was made to cleanup the code and remove redundant lines of code

DivergentEuropeans commented on src/renderer/components/stages/Certificates.tsx:
```diff
@@ -129,7 +129,7 @@ const Certificates = () => {
     stages[STAGE_ID].subStages[SUB_STAGE_ID].isSkipped = status;
     stages[STAGE_ID].isSkipped = status;
-    mapAndSetSkipStatus(SUB_STAGE_ID, status);
+    mapAndSetSubStepSkipStatus(SUB_STAGE_ID, status);
```
If we have `stages` which is state, and we can set both status for the main step and sub step, so looks like we need a `mapAndSetSubStepSkipStatus` for setting local storage, ok sure

but I think
 a) this method should be renamed to like "setAndSaveSubStepSkipStatus"/"setAndStoreSubStepSkipStatus" and "setAndSaveStepSkipStatus"/"setAndStoreStepSkipStatus" because it's doing a storage operation
 b) if these methods already do the setting, why do we need `stages[STAGE_ID].xxxx` stuff at all, shouldn't it be removed
 c) doesn't need to be this PR, but why is `StageProgressStatus.ts` not using a Store?

sakshibobade21 replied:
The function mapAndSetSubStepSkipStatus() is used to update the local storage. stages[STAGE_ID].xxxx modifies the state only for the current session. When the user resumes, we utilize mapAndGetSubStepSkipStatus() to update stage[stage_id].skip.

Throughout the session, stages[STAGE_ID].skip is used to retrieve and update the status. While mapAndGetSubStepSkipStatus() is useful for fetching status, doing so from local storage every time could impact performance negatively.

sakshibobade21 replied:
I agree with you; there is indeed scope for code cleanup and improvement here. We can simplify by calling mapAndSetSubStepSkipStatus(). Within this function, we can set stages[STAGE_ID].skip and update the local storage.

Could we handle the code cleanup in a separate PR?

sakshibobade21 replied:
We also have` mapAndSetStepSkipStatus()` for stages.
I can rename these to `setAndSaveSubStepSkipStatus` but we already have `setSubStageSkipStatus`.
`mapAndSetSubStepSkipStatus` calls `setSubStageSkipStatus` by passing the `stageid` and the `status`.

DivergentEuropeans replied:
> While mapAndGetSubStepSkipStatus() is useful for fetching status, doing so from local storage every time could impact performance negatively.

This is mostly irrelevant because we (should) only update the status after button clicks & operations. If this is negatively affecting performance, that means we are calling it during each render operation which points to an issue elsewhere

> I agree with you; there is indeed scope for code cleanup and improvement here. We can simplify by calling mapAndSetSubStepSkipStatus().

Yes, let's move forward with that and I would prefer to have it done in this PR because you already started some of that duplication removal. So just continue that for everywhere, or do it in a separate PR, but don't begin it & then leave it unfinished (unless you write a detailed TODO)

Rename sounds good 👍 setSubStageSkipStatus and setAndStoreSubStepSkipStatus / setAndSaveSubStepSkipStatus both sound good and make sense, for what they are titled for

DivergentEuropeans replied:
If your setAndSave methods work properly, then the refactor is just a text search and replace so I would prefer it done in the same PR

DivergentEuropeans replied:
Then you would want to check for performance hits, by seeing if the useEffect( )'s or whatever are called after significant necessary updates versus each re-render. It could be out of scope to FIX all the performance hits, but what it means is:
some methods will have
```
stages[STAGE_ID].subStages[SUB_STAGE_ID].isSkipped = status;
    stages[STAGE_ID].isSkipped = status;
```


bla bla bla right, some methods will have
```
setAndStoreSubStepSkipStatus(SUB_STAGE_ID, status);
```

but no methods should have BOTH

sakshibobade21 replied:
We have

 ```
const setStageSkipStatus = (status: boolean) => {
    stages[STAGE_ID].subStages[SUB_STAGE_ID].isSkipped = status;
    stages[STAGE_ID].isSkipped = status;
    mapAndSetSubStepSkipStatus(SUB_STAGE_ID, status);
  }
```

`.isSkipped` is just updating the stages array.
And `mapAndSetSubStepSkipStatus` is updating the local storage.

We can move below code
```
stages[STAGE_ID].subStages[SUB_STAGE_ID].isSkipped = status;
stages[STAGE_ID].isSkipped = status;
```

inside `mapAndSetSubStepSkipStatus`. But I think it is better we keep them separated.

sakshibobade21 replied:
I will try to address the performance improvements in the same PR.

sakshibobade21 replied:
We have
```
const setStageSkipStatus = (status: boolean) => {
   stages[STAGE_ID].subStages[SUB_STAGE_ID].isSkipped = status;
   stages[STAGE_ID].isSkipped = status;
   mapAndSetSubStepSkipStatus(SUB_STAGE_ID, status);
 }
```
`.isSkipped` is just updating the stages array and not the local storage.
And `mapAndSetSubStepSkipStatus` is updating the local storage.

We need to update both.

sakshibobade21 committed:
updating unpax stage

sakshibobade21 committed:
merging

sakshibobade21 committed:
Code cleanup

sakshibobade21 committed:
Optimizing installation

sakshibobade21 committed:
Optimizing the apf auth

sakshibobade21 committed:
optimizing the certs

sakshibobade21 committed:
Updating the vsam stage

sakshibobade21 committed:
optimizing the security stage

sakshibobade21 committed:
updating stcs

sakshibobade21 committed:
Updating the progress slice

sakshibobade21 committed:
Updating the launch config and the network state

sakshibobade21 committed:
Updating the progress slice

sakshibobade21 committed:
Updating the launch config

sakshibobade21 committed:
Complete code cleanup for the unpax stage

sakshibobade21 committed:
minor cleanuo

sakshibobade21 committed:
code fix

sakshibobade21 committed:
Code cleanup

sakshibobade21 committed:
More cleanup

sakshibobade21 committed:
Updating the substages logic

sakshibobade21 committed:
code cleanup for review instllation

sakshibobade21 committed:
updating networking stage

sakshibobade21 committed:
Code renaming

sakshibobade21 commented:
@DivergentEuropeans I addressed all possible code cleanups, both related and unrelated to the stage statuses. I also tried to optimized performance by minimizing calls to functions that interact with local storage, such that these functions are called only once when the component unmounts. Additionally, I renamed the functions we discussed earlier.

sakshibobade21 committed:
Updating the security code

sakshibobade21 committed:
Updating the stcs stage

sakshibobade21 committed:
Updating the certificates stage

sakshibobade21 committed:
Updating the vsam stage

sakshibobade21 committed:
Updating the initapfauth stage

sakshibobade21 committed:
Updating the stepper to update the skip button anabling logic

sakshibobade21 committed:
Updating the unpax stage

sakshibobade21 committed:
More updates to stcs stage

sakshibobade21 committed:
Some more updates to stcs stage

sakshibobade21 committed:
Updating the installation stage

sakshibobade21 committed:
unpax stage bugfix

sakshibobade21 committed:
More changes to the stepper

sakshibobade21 committed:
More changes to all substages

sakshibobade21 committed:
Updates to the unpax stage

sakshibobade21 committed:
Updating the stepper

sakshibobade21 committed:
Updating the edge cases

DivergentEuropeans commented on src/renderer/components/common/Stepper.tsx:
```diff
@@ -240,7 +229,19 @@ export default function HorizontalLinearStepper({stages, initialization}:{stages
+    } else {
+      return true;
+    }
+  }
```
I am a bit confused by this change. So originally, we set `isNextSStepEnabled` based on useAppSelector(selectNextStepEnabled) which comes from the `setNextStepEnabled` state which is a thing where steps have logic to signal if next step is enabled, based on if a certain action has been performed. Ok, sure. 

But now, we are moving away from this to --> check for a `getProgress` which comes from local storage, where a status is being set based on logic within steps upon status completion.

a) So if case 1's state is no longer needed, why don't we remove it? 

b) Why is it called "getProgress"/"setProgress" instead of "get/setStatus" ? 

c) StageProgressStatus.ts in my opinion is a bit long and over-engineered and difficult to read. I think I approved a smaller, earlier version of it many months ago but this was a big mistake. Why are there useless keys there like, 
`let installationTypeKey = 'installation_type';` when this could be 
`INSTALLATION_TYPE_STAGE_LABEL` from Utils.tsx (which was previously just a const in a home/main page somewhere)

You don't need to do anything with c) in this PR

DivergentEuropeans replied:
And btw for c) I'm not talking about localstorage vs redux. That's fine that we used local storage, but the file is still unnecessarily difficult to read

sakshibobade21 replied:
Himani identified an issue in my pull request: the next step was not enabled during the unpax stage when the stage was completed.
While debugging, I discovered that the nextStepButtonDisabled function was being called multiple times, even when the stage status remained unchanged. This caused the data to be fetched from local storage repeatedly. Although this didn't significantly impact performance, it would be more efficient to use the setNextStepEnabled function instead.
Therefore, I updated the code to use setNextStepEnabled, along with some modifications.

DivergentEuropeans commented on src/renderer/components/stages/Certificates.tsx:
```diff
@@ -144,10 +151,15 @@ const Certificates = () => {
+  const dispatchActions = (status: boolean) => {
+    dispatch(setCertificateStatus(status));
+    dispatch(setInitializationStatus(isInitializationStageComplete()));
+    dispatch(setNextStepEnabled(status));
```
If we are now using "Status" as the signal that tells us if next step is enabled, then all of these `setNextStepEnabled` now become useless no?

sakshibobade21 replied:
Himani identified an issue in my pull request: the next step was not enabled during the unpax stage when the stage was completed.
While debugging, I discovered that the nextStepButtonDisabled function was being called multiple times, even when the stage status remained unchanged. This caused the data to be fetched from local storage repeatedly. Although this didn't significantly impact performance, it would be more efficient to use the setNextStepEnabled function instead.
Therefore, I updated the code to use setNextStepEnabled, along with some modifications.

sakshibobade21 replied:
Addressed - https://github.com/zowe/zen/pull/193#discussion_r1668324613

sakshibobade21 committed:
Restoring the status

sakshibobade21 committed:
Updating the init apf auth stage

sakshibobade21 committed:
Updating the init mvs stage

sakshibobade21 committed:
Updating all the substages

sakshibobade21 committed:
Updating the pax stage

sakshibobade21 committed:
Updating the review installation stage

sakshibobade21 committed:
Updating the unpax

sakshibobade21 committed:
Merging with the dry ru code

sakshibobade21 committed:
Updating the indtallation type stage

sakshibobade21 committed:
Merging with staging

sakshibobade21 committed:
Merge remote-tracking branch 'origin' into regression/vsam-status

skurnevich reviewed (approved):
So in the current state we do not distinguish between failed and skipped steps?  

Also, for instance if Certificates step fails - it shows job output as error, but then we skip it and go to vsam (which should be completely optional) and it fails too - it will show error in the header, and the job log stays from the previous failure in Certificates. It looks inconsistent and misleading. 

I do not mind merging this though, as it doesn't break things and in order to continue the development process.

sakshibobade21 commented:
@skurnevich Thank you for reviewing the PR.

At present, in ZEN, we do not differentiate between failed and skipped steps. This could be a potential enhancement to consider for future improvements.

Regarding your comment about the certificate stage, I think we should address that in a separate PR since it originates from the staging. I will create a ticket for this.

Thanls.

skurnevich commented:
Hi Sakshi, that is fine with me, please resolve Lenny's changes request and we can merge this PR

sakshibobade21 commented:
@skurnevich thanks. And the changes requested by Lenny have already been addressed.

Also, I do not have the merge rights for this repository. Could you please merge it if you are ok with it? This PR will serve as a foundation for future PRs in ZEN.

skurnevich commented:
Hi @sakshibobade21 , i'll merge this PR, but problem is the github still has "Changes requested" as blocker for merge there.
<img width="907" alt="Screenshot 2024-08-06 at 11 11 15" src="https://github.com/user-attachments/assets/29cbb913-aef3-4c37-bb38-88999dd6db7f">
As far as i can see the requests were addressed, and I can merge it bypassing the protection if @DivergentEuropeans  does not mind. 
But the proper way to resolve this is to re-request review, 
<img width="399" alt="Screenshot 2024-08-06 at 11 25 24" src="https://github.com/user-attachments/assets/a50ddd5a-d75c-47df-ab53-86294057591b">
Github sometimes acts weird with permissions, so can you please check if you have this option in the top right corner first?

sakshibobade21 commented:
@skurnevich I re-requested the review now. Hope that works.
Thank you.

sakshibobade21 committed:
Merging with staging

DivergentEuropeans commented:
Thanks, dismissed stale review @sakshibobade21 @skurnevich
